Duplicati/Library/Main/Operation/ListFilesHandler.cs

```diff
@@ -203,9 +204,20 @@ orderby n.Key
             return filelistFilter(parsedlist).ToList();
         }
 
-        public static IEnumerable<Library.Interface.IListResultFileset> CreateResultSequence(IEnumerable<KeyValuePair<long, Volumes.IParsedVolume>> filteredList)
+        private static IEnumerable<IListResultFileset> CreateResultSequence(IEnumerable<KeyValuePair<long, IParsedVolume>> filteredList, BackendManager backendManager, Options options)
         {
-            return (from n in filteredList select (Library.Interface.IListResultFileset)(new ListResultFileset(n.Key, BackupType.PARTIAL_BACKUP, n.Value.Time.ToLocalTime(), -1, -1))).ToArray();
+            List<IListResultFileset> list = new List<IListResultFileset>();
+            foreach (KeyValuePair<long, IParsedVolume> entry in filteredList)

```

Comment:
Can you add some comments to this function what is happening?

Reply:
Good idea.  I added a comment in revision fa3ffb75400f4e954e8b1f62ab4bc050ea9d330a to explain why the `dlist` files are being downloaded, as well as the resulting impact on performance.

Reply:
Revision afecf891dab745a7f06e74e9b47687e12a09e3a1 adds a test for the `--keep-time` retention logic.  I'll try and write a test for the `--keep-versions` logic soon.